arxiv/ops/fastly_log_ingest/app.py

```diff
@@ -0,0 +1,145 @@
+    while RUN:
+        batch: List[WorkItem] = []
+        with cv:
+            cv.wait_for(lambda: shared_messages, timeout=2.0)  # timeout to allow exit or periodic message
+            now = perf_counter()
+            msg_count = len(shared_messages)
+            if msg_count and (msg_count > MAX_PER_BATCH or (now - shared_messages[0].arrival_time) > SEND_PERIOD):
+                batch.extend([shared_messages.pop() for _ in range(min([msg_count, MAX_PER_BATCH]))])
+            pass
+        if batch:
+            logger.debug(f"Got batch of %d pub/sub messages", len(batch))
+            logs_to_gcp(log_client, [item.to_log_entry() for item in batch])
+            logger.debug("Finished sending entries to gcp logging.")
+            for item in batch:
+                item.message.ack()
+                logger.debug("ack pub/sub message %s", item.message.message_id)
+            log_write_rate.event(quantity=len(batch))
+        now = perf_counter()
+        if now - last_info_msg >= INFO_PERIOD:
+            last_info_msg = now
+            logger.info(log_write_rate.rate_msg())
+    logger.info(log_write_rate.rate_msg())  # on exit
+
+
+def _pubsub_callback(message: Message) -> None:
+    global last_pubsub_info
+    arrival = perf_counter()
+    with cv:
+        shared_messages.append(WorkItem(message, arrival))
+        logger.debug("got message %s and added to queue", message.message_id)
+        msg_rate.event()
+        now = perf_counter()
+        if now - last_pubsub_info > INFO_PERIOD:
+            last_pubsub_info = now
+            logger.info(msg_rate.rate_msg())
+
+
+if __name__ == "__main__":
+    credentials, project = google.auth.default()
+    logger.info("GOOGLE_APPLICATION_CREDENTIALS envvar:"
+                f"{os.environ.get('GOOGLE_APPLICATION_CREDENTIALS', '(not set)')}")
+    _log_credentials(credentials)
+    logger.info("about to start logging threads")
+    threads = []
+    for _ in range(0, THREADS):
+        threads.append(Thread(target=_logging_thread))
+    [t.start() for t in threads]
+    logger.info(f"Started {len(threads)} log writer threads.")
+
+    """>>>>>>>> Setup of streaming pull from Pub/Sub <<<<<<<<"""
+
+    subscriber = pubsub_v1.SubscriberClient()
+    # The `subscription_path` method creates a fully qualified identifier
+    # in the form `projects/{project_id}/subscriptions/{subscription_id}`
+    subscription_path = subscriber.subscription_path(PROJECT_ID, SUBSCRIPTION_ID)
+
+    streaming_pull_future = subscriber.subscribe(subscription_path, callback=_pubsub_callback)
+    logger.info(f"Listening for pub/sub messages on {subscription_path}..\n")
+    with subscriber:
+        try:
+            streaming_pull_future.result(timeout=None)
+        except Exception as ex:
+            logger.error("Error with pub/sub subscription", exc_info=ex)
+            if isinstance(ex, PermissionDenied):
+                _log_credentials(credentials)
+            RUN = False
```

Comment:
Make RUN = False as the first of exception block. In some unfortunate event where _log_cred() to raise, RUN may remain True. (defensive programming)